src/cargo/util/network/mod.rs

```diff
@@ -20,20 +20,51 @@ impl<T> PollExt<T> for Poll<T> {
     }
 }
 
-// When dynamically linked against libcurl, we want to ignore some failures
-// when using old versions that don't support certain features.
+/// When dynamically linked against libcurl, we want to ignore some failures
+/// when using old versions that don't support certain features.
 #[macro_export]
 macro_rules! try_old_curl {
     ($e:expr, $msg:expr) => {
         let result = $e;
         if cfg!(target_os = "macos") {
             if let Err(e) = result {
-                warn!("ignoring libcurl {} error: {}", $msg, e);
+                ::log::warn!("ignoring libcurl {} error: {}", $msg, e);
             }
         } else {
+            use ::anyhow::Context;
             result.with_context(|| {
-                anyhow::format_err!("failed to enable {}, is curl not built right?", $msg)
+                ::anyhow::format_err!("failed to enable {}, is curl not built right?", $msg)
             })?;
         }
     };
 }
+
+/// Enable HTTP/2 and pipewait to be used as it'll allow true multiplexing
+/// which makes downloads much faster.
+///
+/// Currently Cargo requests the `http2` feature of the `curl` crate which
+/// means it should always be built in. On OSX, however, we ship cargo still
+/// linked against the system libcurl. Building curl with ALPN support for
+/// HTTP/2 requires newer versions of OSX (the SecureTransport API) than we
+/// want to ship Cargo for. By linking Cargo against the system libcurl then
+/// older curl installations won't use HTTP/2 but newer ones will. All that to
+/// basically say we ignore errors here on OSX, but consider this a fatal error
+/// to not activate HTTP/2 on all other platforms.
+///
+/// `pipewait` is an option which indicates that if there's a bunch of parallel
+/// requests to the same host they all wait until the pipelining status of the
+/// host is known. This means that we won't initiate dozens of connections but
+/// rather only one. Once the main one is opened we realized that pipelining is
+/// possible and multiplexing is possible. All in all this reduces the number
+/// of connections down to a more manageable state.
+#[macro_export]
+macro_rules! try_old_curl_http2_pipewait {
```

Comment:
Is there a reason you did this as a macro?

Reply:
It was a macro so I followed. It could be quite annoying to write a shared function for this I guess, but I can try if we don't want macros.

Reply:
I think you mean it is wrapping / extending an existing macro?

I personally lean towards only using macros when there is a clear need for it.  On the surface, it seems like both could be functions *but* I can see where you are coming from in how this fits with improving within what exists, even if there are potential further improvements.